Instructor: Help page: Documentation search feature #8950

mg14777 opened the pull request:
Fixes #8950

**PR Checklist**
- [x] Add UI elements
- [x] Add JS logic
- [ ] Add Enhancements
- [ ] Add Tests

Work ongoing

mg14777 commented:
@damithc @wenmogu Highlighting keyword matches now and built a pagination system. Please have a look

Link to staging server: https://teammates-mudit.appspot.com/instructorHelp.jsp

![search_pagination_highlight](https://user-images.githubusercontent.com/11634699/42667162-70c695d6-8641-11e8-8acd-7b49bc20a121.gif)

damithc commented:
Nice. 
* Try to use the same yellow highlighting we use elsewhere in TEAMMATES, for consistency.
* Does the search covers panel text, not just panel headings? If not, it's better if it did. And if they did, it is nice if we can highlight those keywords too.

craaaa commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,224 @@ function bindPanelCollapseLinksAndAnchor() {
+                }
+                return elasticlunr.stemmer(lowerCaseWord);
+            });
+            subTopicTags = $.grep(subTopicTags, (v) => {
```
Choose a better variable name than `v`?

craaaa commented on src/main/webapp/instructorHelp.jsp:
```diff
@@ -31,12 +31,39 @@
+    <div class="row">
+      <div class="col-sm-6 col-sm-offset-3">
+        <div class="input-group input-group-lg">
+            <input type="text" class="form-control" placeholder="How can we help ?" id="searchQuery" autocomplete="off">
```
Remove the space before the question mark

craaaa commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -1,3 +1,12 @@
+/* global elasticlunr:true */
+
+const questionMap = {};
+const numResultsPerPage = 5;
```
Any specific reason why you chose 5 results per page?

mg14777 replied:
i was trying to test if the pagination scales well so used small number of results per page. For the final version i was thinking 10 ? Some panels are quite long particularly `Questions` category in which case 10 might be overwhelming ?

craaaa commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,224 @@ function bindPanelCollapseLinksAndAnchor() {
+
+    // highlight matches for keyword in the search results
+    function highlightKeyword(keyword) {
+        const highlightRegex = new RegExp(keyword, 'g');
```
This regex appears to be case-sensitive

mg14777 replied:
Good point. Fixed it.

craaaa commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,224 @@ function bindPanelCollapseLinksAndAnchor() {
+        $('#searchResults').children().each(function () {
+            // highlight heading matches for query
+            $(this).find('.panel-heading').html(function () {
+                return $(this).html().replace(highlightRegex, `<span class='text-bold color-positive'>${keyword}</span>`);
```
When you change the styling, ensure that it contrasts well with both regular text and text in examples (e.g. the following is a negative example)
![screen shot 2018-07-14 at 2 34 19 am](https://user-images.githubusercontent.com/12183503/42703054-af8fbe50-870e-11e8-9a59-6a9578aff879.png)

craaaa commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,224 @@ function bindPanelCollapseLinksAndAnchor() {
+                if (lowerCaseWord.length < 3) {
+                    return '';
+                }
+                return elasticlunr.stemmer(lowerCaseWord);
```
Be careful with stemming; while it's useful for searching, if the search matches only the stemmed version but not the original version, there will be no visual evidence of the match through highlighting so results might seem inconsistent.

e.g. search for `teams` -- search (correctly) retrieves questions that deal with `team`, but since only exact matches are highlighted, some questions are deemed relevant without displaying any highlights at all. 

For consistency, I would prefer if either we highlighted words whose stemmed forms match (e.g. highlight `team` even though search term was `teams` and vice versa), or don't highlight any words.

mg14777 replied:
I agree with your point. I was experimenting with this yesterday and basically if I remove stemming altogether from the processing then only those results that contain the exact match are returned in majority of cases. This solves the highlight problem but we miss out on potentially relevant questions whose stems match.

If we highlight the stems instead we might get wierd cases due to the way stemmer works. For example, the stem for "course" and "courses" is "cours" which might not make sense if highlighted ?

I think we should keep highlighting since it helps in picking relevant information in large panels.

mg14777 commented:
Made appropriate changes, added tests and as pointed by @wenmogu i have added responsiveness to pagination system. Staging server updated.
https://teammates-mudit.appspot.com/instructorHelp.jsp

Please let me know if it looks good 

![search_version_2](https://user-images.githubusercontent.com/11634699/43136090-82d75204-8f3f-11e8-9753-a81c070da6d1.gif)

mg14777 commented:
Fixed the overflowing issue by changing the `Previous` and `Next` buttons and reducing the size of the toolbar altogether. Also highlighting the current page now.

Staging server updated.

![image](https://user-images.githubusercontent.com/11634699/43144024-0ff1b97c-8f54-11e8-94e9-8944eebe557a.png)

mg14777 commented:
@wenmogu @craaaa Just a reminder for review.

mg14777 commented:
@craaaa Reminder for review

mg14777 commented:
@craaaa Reminder for review

Staging server: [https://teammates-mudit.appspot.com](https://teammates-mudit.appspot.com)

Final changes:

![search_final_1](https://user-images.githubusercontent.com/11634699/43678571-9baf05d4-980d-11e8-88ba-3d401006cad4.gif)

craaaa commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,319 @@ function bindPanelCollapseLinksAndAnchor() {
     bindPanelCollapseLinksAndAnchor();
+    prepareQuestionsForSearch();
+    bindEnterKeyForSearchBox();
+    resetSearch();
```
Rename this function? It doesn't quite make sense to reset search if the search hasn't happened yet. Perhaps something like `displayDefaultView` (maybe not this exactly) might be more appropriate.

mg14777 replied:
Sure. How about `displayDefaultAndClearSearchBox` or is that too long ?

craaaa replied:
That's good

mg14777 commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,319 @@ function bindPanelCollapseLinksAndAnchor() {
+    renderPage(1);
+    $('#pagingDivider').show();
+    $('#searchResults').show();
+    $('#searchMetaData').text(`${results.length} results found for '${query}'`);
```
Can you give an example of the queries you tried ? The longest i tried was `What should I do if a student says he/she cannot submit an evaluation due to a technical glitch?` and i got this
Is this what you meant ?

![image](https://user-images.githubusercontent.com/11634699/43680082-5ed64e7a-982a-11e8-8750-fafc273b70b1.png)

mg14777 commented:
@damithc Ready for final review and merging

damithc commented:
As this PR has a lot of JS code, I'd like @xpdavid to do the final code quality review for this PR.

xpdavid commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,319 @@ function bindPanelCollapseLinksAndAnchor() {
+                const htmlId = question.attr('id');
+                const tagsSeenSoFar = {};
+
+                let questionTags = subTopicTags.concat(htmlId.split('-'));
```
Merge two arrays?
`jQuery.merge()`?

Why the tags should care about the htmlID?

mg14777 replied:
question ids provide relevant keywords regarding the question which can added as tags eg. `student-view-responses`

xpdavid replied:
Why not extract keywords from question title rather than question id?

mg14777 replied:
we'll need to process and tokenize it since it contains lots of irrelevant words and prepositions. This info is readily available from the ids so i thought just use it as is

xpdavid replied:
Okay.

mg14777 commented on src/main/java/teammates/common/util/FrontEndLibrary.java:
```diff
@@ -44,10 +46,12 @@ private FrontEndLibrary() {
         BLANKET = getLibrarySource("blanket", "dist/qunit/blanket.min.js");
         BOOTBOX = getLibrarySource("bootbox", "bootbox.min.js");
         BOOTSTRAP_CSS = getLibrarySource("bootstrap", "dist/css/bootstrap.min.css");
+        BOOTSTRAP_FORMS_LESS = getLibrarySource("bootstrap", "less/forms.less");
```
it was being used by highlighting initially but after I fixed some bugs it wasn't necessary anymore. Thanks for pointing out.

mg14777 commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,319 @@ function bindPanelCollapseLinksAndAnchor() {
+
+    // highlight matches for keyword in the search results
+    function highlightKeyword(keyword) {
+        const highlightRegex = new RegExp(keyword, 'gi');
```
yeah valid point. If I use `page\` then i get invalid regexp. What should we use for escaping ?

xpdavid replied:
If you use a library, it will take care for you :D. https://markjs.io/

mg14777 commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,319 @@ function bindPanelCollapseLinksAndAnchor() {
+    });
+
+    // highlight matches for keyword in the search results
+    function highlightKeyword(keyword) {
```
well there's mark.js. We could use that ?

mg14777 replied:
No I mean we should. Was just confirming. No point reinventing the wheel.

mg14777 commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,319 @@ function bindPanelCollapseLinksAndAnchor() {
+    let questionCount = 0;
+
+    function addQuestionsForTopic(topic) {
+        const topicContentElement = $(topic).next();
```
currently we're using `contentHolder` as id of all topics not sure why ? If we change that to a class will the class be empty since there's no CSS rule for it atm

xpdavid replied:
You can use the class without define it. It's fine.

mg14777 commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,319 @@ function bindPanelCollapseLinksAndAnchor() {
+        index = elasticlunr();
+        index.addField('title');
+        index.addField('body');
+        index.addField('tags');
```
i guess we could. Although currently I am using topics to refer to top level classifications as shown on the page eg. Students, Sessions, Courses and Questions. Tags refer to refined tokens describing each question. For example, subtopic headings are part of tags like `Enrolling Students in Courses` which is tokenized to enroll, student etc. Similarly I am parsing ids of each question to get tags.

xpdavid replied:
Make sense. We can use `tags`.

mg14777 commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,319 @@ function bindPanelCollapseLinksAndAnchor() {
+    // add relevant panels to search results
+    $.each(results, (idx, result) => {
+        const { htmlId } = questionMap[result.ref];
+        const questionPanel = $(`#${htmlId}`)[0];
```
actually i just tested it and i do need it since i need to obtain the outerhtml and for that I need the native DOM element. Fastest way of getting it is using indexing as shown below:
https://learn.jquery.com/using-jquery-core/faq/how-do-i-pull-a-native-dom-element-from-a-jquery-object/
If you have a better suggestion please share

xpdavid replied:
Can use jQuery to do the manipulation rather than the native API? Though I have no objection to add [0].

mg14777 commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -209,12 +205,30 @@ function displayDefaultAndClearSearchBox() {
     $('#topics').show();
 }
 
+function bindPagingButtons() {
```
@xpdavid does this look good ?

xpdavid commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -209,12 +205,30 @@ function displayDefaultAndClearSearchBox() {
+    }
+    $('#pagingControls').find('button').each((idx, pageButton) => {
+        if ($(pageButton).text() !== '...') {
+            if ($(pageButton).attr('id') === 'prevPage') {
```
Use `data` attribute instead of id?

mg14777 commented:
@xpdavid Ready for review

mg14777 commented:
Can we merge this today if everything looks good ?

xpdavid commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,301 @@ function bindPanelCollapseLinksAndAnchor() {
+
+    // highlight matches for keyword in the search results
+    function highlightKeyword(keyword) {
+        $('#searchResults').find('.panel.panel-default').each(function () {
```
Shall we use `() => {}` to define function instead just for consistency?

mg14777 commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,301 @@ function bindPanelCollapseLinksAndAnchor() {
+                const tagsSeenSoFar = {};
+
+                let questionTags = [];
+                $.merge($.merge(questionTags, subTopicTags), htmlId.split('-'));
```
No if I am not wrong `$.merge(subTopicTags, htmlId.split('-'))` will modify `subTopicTags` and it needs to stay constant for all `subTopicQuestions`. Also `merge` doesn't eliminate duplicates.

xpdavid replied:
Seems `$.extend(arr1,arr2)` could?

Just try to use existing function in JQuery to write more concise code.

xpdavid replied:
As we are using ES6 🎉 
```
var a = [34, 35, 45, 48, 49];
var b = [48, 55];
var union = [...new Set([...a, ...b])];
```

mg14777 commented on src/test/java/teammates/test/cases/browsertests/InstructorHelpPageUiTest.java:
```diff
@@ -0,0 +1,82 @@
+        helpPage.inputSearchQuery(searchContent);
+        helpPage.clickSearchButton();
+        String searchResultsMultipleKeywords = helpPage.getSearchResults();
+        helpPage.verifyHtmlMainContent("/instructorHelpPageSearchMultipleKeywords.html");
```
> I prefer to just verify the searchResult box

You mean check that no. of results returned is same in both cases ?

xpdavid replied:
Not no. of results but the whole content in the results elements.

mg14777 commented on src/main/webapp/dev/js/main/instructorHelp.js:
```diff
@@ -25,7 +34,301 @@ function bindPanelCollapseLinksAndAnchor() {
+ * re-initializes the pagination toolbar to highlight the new page.
+ * For Previous and Next buttons the function is simply called recursively
+ * after incrementing the page counter.
+ * @param page current page to be rendered
```
you mean space between `page` and the comment `current page ...` ? isn't there one already ?

xpdavid replied:
Opps. Should be a empty line.

```java
......
* after incrementing the page counter.
*
* @param page current page to be rendered
......
```

xpdavid commented:
@mg14777 Let's get this merged by tmr evening.

mg14777 commented on src/test/java/teammates/test/cases/browsertests/InstructorHelpPageUiTest.java:
```diff
@@ -0,0 +1,82 @@
+        helpPage.clearSearchBox();
+        helpPage.inputSearchQuery(searchContent);
+        helpPage.clickSearchButton();
+        assertEquals(searchResultsMultipleKeywords, helpPage.getSearchResults());
```
I was just thinking is a file needed now since I am just comparing the search meta data which is a sentence. 
If I compare it against a file I will need to lowercase the content first since we need to ignore the case ?

xpdavid replied:
The file only contain part of the page rather than the whole page.

mg14777 replied:
Ok makes sense. However, another question I had was how would I reuse the file for case insensitive  test ? Essentially I'm trying to check that `43 results found for enroll students` is same as `43 results found for ENROLL STUDENTS` but ignoring the case of the query. I direct comparison against the file wouldn't work right.

xpdavid replied:
Good questions... I thought your result DIV doesn't contain the information but pure results. NVM, you can keep the current way and prepare your staging server.

mg14777 replied:
Apologies I think I'm still not being very clear :sweat_smile: 
Basically there are 2 divs `searchMetaData` and `searchResults`. 

`searchMetaData` describes the number of results returned like i mentioned above (`43 results found for enroll students`)
`searchResults` is a list of pure result panels.

When you said we should check that no. of results returned is same I thought you meant `searchMetaData` but I think you actually mean the latter. In that case what you're saying will work since `searchResults` contains pure results. 

Sorry for the confusion. I'll make the changes

mg14777 commented on src/test/java/teammates/test/cases/browsertests/InstructorHelpPageUiTest.java:
```diff
@@ -46,24 +48,23 @@ private void testSearch() throws Exception {
         helpPage.inputSearchQuery(searchContent);
         helpPage.clickSearchButton();
-        helpPage.verifyHtmlMainContent("/instructorHelpPageSearchSingleKeyword.html");
+        helpPage.verifyHtmlPart(SEARCH_RESULTS, "/instructorHelpPageSearchSingleKeyword.html");
```
@xpdavid does this look good ?

mg14777 commented:
@xpdavid Ready for review.

Staging server: https://teammates-mudit.appspot.com

xpdavid commented:
@mg14777 The `InstructorHelpPageUiTest.java` is failing in the [AppVeyor](https://ci.appveyor.com/project/damithc/teammates/build/1.0.1112) build. Can you help to investigate the reason?

dalessr commented:
@xpdavid Sorry my bad :( Added something else (related to `courses`) in `InstructorHelpPage` at commit https://github.com/TEAMMATES/teammates/commit/f49fc70057c1706db58d06c2f60ff348bb9c9963. Will go investigate later.
